cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,93 @@
+    
+    temp.resize(n, stream);
+    kernelA<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    kernelB<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    temp.release(stream);
+}
+```
+## Using thrust
+To ensure that thrust algorithms allocate temporary memory via the provided device memory allocator the `ML::thrustAllocatorAdapter` available in `allocatorAdapter.hpp` should be used with the `thrust::cuda::par` execution policy:
+```cpp
+void foo( ML::cumlHandle* handle, .. )
+{
+    ML::thrustAllocatorAdapter alloc( handle->getDeviceAllocator(), handle->getStream() );
+    auto execution_policy = thrust::cuda::par(alloc).on(handle->getStream());
+    thrust::for_each(execution_policy, ... );
+}
+```
+# Asynchronous operations and stream ordering
+All ML algorithms should be as asynchronous as possible avoiding the use of the default stream (aka as NULL or `0` stream). If an implementation only requires a single CUDA Stream the stream from `ML::cumlHandle` should be used:
+```cpp
+void foo( ML::cumlHandle* handle, ...)
+{
+    cudaStream_t stream = handle->getStream();
+}
+```
+In case multiple streams are needed, e.g. to manage a pipeline, the internal streams available in `ML::cumlHandle_impl` should be used (see [CUDA Resources](# CUDA Resources)). If multiple streams are used all operations still need to be ordered according to `ML::cumlHandle::getStream()`, i.e. before any operation in any of the internal CUDA streams is started all previous work in `ML::cumlHandle: getStream()` needs to have completed and any work enqueued in `ML::cumlHandle::getStream()` after an cuML function returns should not start before all work enqueued in the internal stream has complete. This can be ensured by introducing inter stream dependencies with CUDA events and `cudaStreamWaitEvent`. For convenience  the header `cumlHandle.hpp` provide the class `ML::detail::streamSyncer` which lets all `ML::cumlHandle_impl` internal CUDA streams wait on `ML::cumlHandle::getStream()` in its constructor and in its destructor and lets `ML::cumlHandle::getStream()` wait on all work enqueued in the `ML::cumlHandle_impl` internal CUDA streams. E.g. a usage like this:
+```cpp
+void cumlAlgo( ML::cumlHandle* handle, ...)
+{
+    ML::detail::streamSyncer _(handle);
+}
+```
+ensures the stream ordering behavior described above.
+
+## Using thrust
+To ensure that thrust algorithms are executed in the intended stream the `thrust::cuda::par` execution policy should be used (described in the section Device and Host memory allocations/Using thrust).
+
+# CUDA Resources
+Implementations of ML algorithms should not create reusable resources themselves. Instead they should use the existing one in `ML::cumlHandle_impl `. This allows to avoid constant creating and recreation of reusable resources such as CUDA streams, CUDA events or library handles. Please file a feature request in case a resource handle is missing in `ML::cumlHandle_impl `.
```

Comment:
The basic idea is that the handle should own a pool of streams. So that we avoid constant creation and destruction of streams. I have not spend too much thought if this should just be a pool which can grow on demand with an interface like
```cpp
cudaStream_t stream = handle->getImpl()->getStream();
//use stream
handle->getImpl()->releaseStream( stream );
```
of if a handle simply has a fixed set of stream which are accessed like this
```cpp
cudaStream_t stream = handle->getImpl()->getStream(stream_id);
```
I think I would current prefer the former. The latter has the advantage that we would not need a releaseStream and a another smart RAII wrapper object that takes care that a stream is returned to the pool when going out of scope. What do you think?

Reply:
> cudaStream_t stream = handle->getImpl()->getInternalStream(stream_id);

So there would need to be some way to query what the available `stream_id`s are, yes? 

> The fixed set is required to make `ML::detail::streamSyncer` work and over a stream pool

Does it need to be a fixed set? Couldn't you maintain a dynamically sized `vector` or `queue` of streams? It could grow as needed (but probably never shrink?) in response to additional requests for streams. 

`ML::detail::streamSyncer` could then iterate the streams in the pool at construction/destruction to synchronize accordingly. Maybe I'm missing something, but I think this should work even if the size of the pool isn't fixed.

Reply:
> > cudaStream_t stream = handle->getImpl()->getInternalStream(stream_id);
> 
> So there would need to be some way to query what the available `stream_id`s are, yes?

Yes that is needed.

> > The fixed set is required to make `ML::detail::streamSyncer` work and over a stream pool
> 
> Does it need to be a fixed set? Couldn't you maintain a dynamically sized `vector` or `queue` of streams? It could grow as needed (but probably never shrink?) in response to additional requests for streams.
> 
> `ML::detail::streamSyncer` could then iterate the streams in the pool at construction/destruction to synchronize accordingly. Maybe I'm missing something, but I think this should work even if the size of the pool isn't fixed.

What I had in mind is that stream created after the constructor of `ML::detail::streamSyncer` was called would not wait for the begin event recorded in the user stream. Would need to clarify this, but I guess stream creation implicitly synchronizing so this probably can't happen. However that is then the reason why we want to avoid dynamic stream creation. I just pushed a proposal adding `cudaStream_t getInternalStream( int sid ) const` and `int getNumInternalStreams() const` to `cumlHandle_impl` with a fixed number of internal streams.